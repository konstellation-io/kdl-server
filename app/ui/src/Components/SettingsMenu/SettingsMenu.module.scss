@import 'Styles/colors';
@import 'Styles/mixins';
@import 'Styles/borders';
@import 'Styles/shadows';

.settings > div > div {
  background-color: transparent !important;
}

.settingButton {
  background-color: palette(base, 400);
  border-left: 3px solid transparent;
  transition:
    background-color ease 0.3s,
    border-left-color ease 0.3s,
    color ease 0.2s, opacity ease 0.3s;

  &:hover {
    border-left-color: palette(highlight);
    color: palette(highlight);
    text-shadow: 0 0 16px  palette(highlight);
  }
}

.logoutButton {
  background-color: palette(base, 400);
  border-left: 3px solid transparent;
  transition:
    background-color ease 0.3s,
    border-left-color ease 0.3s,
    color ease 0.2s, opacity ease 0.3s;
  color: rgb(204, 110, 110);
  &:hover {
    border-left-color: rgb(204, 110, 110);
    color: red;
    text-shadow: 0 0 16px  rgb(204, 110, 110);
  }
}

.logoutMessage {
  @include font-body;
  color: font-color(lowlight);
  white-space: initial;
  text-align: left;
}

.separator {
  @include font-caption;
  color: font-color(regular);

  cursor: default;
  background-color: palette(base, 400);
  height: 5 * $grid-unit;
  line-height: 5 * $grid-unit;
  padding-left: 3 * $grid-unit;

  &:hover {
    background-color: palette(base, 400);
    color: font-color(regular);
  }
}

.releaseVersion {
  @extend .separator;

  @include font-small;
  text-align: right;

  padding-right: 2 * $grid-unit;
  height: 4 * $grid-unit;
  line-height: 4 * $grid-unit;
  border-top: 2px solid palette(base, 500);
}
