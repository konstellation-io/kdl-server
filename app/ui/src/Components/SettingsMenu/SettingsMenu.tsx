import { Button, BUTTON_ALIGN, CustomOptionProps, Select, SelectTheme, SpinnerCircular } from 'kwc';
import ROUTE from 'Constants/routes';
import React, { memo } from 'react';
import { useHistory } from 'react-router-dom';

import { GetMe } from 'Graphql/queries/types/GetMe';
import KeyIcon from '@material-ui/icons/VpnKey';
import DescriptionIcon from '@material-ui/icons/Description';
import ExitToApp from '@material-ui/icons/ExitToApp';
import styles from './SettingsMenu.module.scss';
import { useQuery, useReactiveVar } from '@apollo/client';
import { CONFIG } from 'index';

import GetMeQuery from 'Graphql/queries/getMe';
import { runningRuntime } from 'Graphql/client/cache';

const UserSettingsSeparator = ({ label }: CustomOptionProps) => (
  <div className={styles.separator}>{label.toUpperCase()}</div>
);

const ReleaseVersion = ({ label }: CustomOptionProps) => <div className={styles.releaseVersion}>{label}</div>;

function SettingsMenu() {
  const history = useHistory();
  const { data, loading } = useQuery<GetMe>(GetMeQuery);
  const runtimeRunning = useReactiveVar(runningRuntime);

  if (loading) return <SpinnerCircular />;

  function goToUserSSHKeys() {
    history.push(ROUTE.USER_SSH_KEY);
  }

  function goToUserKubeconfig() {
    history.push(ROUTE.USER_KUBECONFIG);
  }

  function goToLogout() {
    window.location.href = ROUTE.LOGOUT;
  }

  function SSHKeyButton() {
    return (
      <Button
        label="SSH key"
        key="SSH key"
        onClick={goToUserSSHKeys}
        Icon={KeyIcon}
        className={styles.settingButton}
        align={BUTTON_ALIGN.LEFT}
      />
    );
  }

  function kubeconfigButton() {
    if (!data?.me.isKubeconfigEnabled || !runtimeRunning) {
      return (
        <Button
          label="Kubeconfig"
          key="Kubeconfig"
          disabled
          Icon={DescriptionIcon}
          className={styles.settingButtonDisabled}
          align={BUTTON_ALIGN.LEFT}
        />
    )};

    return (
      <Button
        label="Kubeconfig"
        key="Kubeconfig"
        onClick={goToUserKubeconfig}
        Icon={DescriptionIcon}
        className={styles.settingButton}
        align={BUTTON_ALIGN.LEFT}
      />
    );
  }

  function logoutButton() {
    return (
      <Button
        label="Logout"
        key="logout"
        onClick={goToLogout}
        Icon={ExitToApp}
        className={styles.logoutButton}
        align={BUTTON_ALIGN.LEFT}
      />
    );
  }

  const optionToButton = {
    'user settings': UserSettingsSeparator,
    'ssh key': SSHKeyButton,
    kubeconfig: kubeconfigButton,
    'log out': logoutButton,
    [CONFIG.RELEASE_VERSION]: ReleaseVersion,
  };

  return (
    <div data-testid="settingsCrumb">
      <Select
        label=""
        placeholder={data?.me?.email}
        options={Object.keys(optionToButton)}
        theme={SelectTheme.DARK}
        CustomOptions={optionToButton}
        className={styles.settings}
        showSelectAllOption={false}
        shouldSort={false}
        hideError
      />
    </div>
  );
}

export default memo(SettingsMenu);
