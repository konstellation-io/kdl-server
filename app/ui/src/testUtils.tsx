import * as React from 'react';
import { ApolloClient, ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { mount } from 'enzyme';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function apolloRender(component: JSX.Element, client: ApolloClient<any>) {
  render(<ApolloProvider client={client}>{component}</ApolloProvider>);
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export const HookWrapper = ({ hook }: any) => <div hook={hook()} />; // eslint-disable-line @typescript-eslint/no-explicit-any
export async function apolloHookRender(
  // eslint-disable-next-line @typescript-eslint/ban-types
  hook: Function,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  client: ApolloClient<any>,
) {
  const wrapper = mount(
    <ApolloProvider client={client}>
      <HookWrapper hook={hook} />
    </ApolloProvider>,
  );

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return wrapper.find('div').props().hook;
}

export const loadingHandler = () =>
  new Promise(() => {
    // This is intentional
  });
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const dataHandler = (data: any) => () => Promise.resolve({ data });
export const errorHandler = (error: string) => () => Promise.reject({ error });

export function getSnapshot() {
  return document.querySelector('body');
}

export const ERROR_MESSAGE = 'Something went wrong, refresh the page or try again later.';
