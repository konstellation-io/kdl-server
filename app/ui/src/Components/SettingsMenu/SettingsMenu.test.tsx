import * as React from 'react';
import SettingsMenu from './SettingsMenu';
import { createMockClient } from 'mock-apollo-client';
import { screen, fireEvent, cleanup } from '@testing-library/react';
import GetMeQuery from 'Graphql/queries/getMe';

import data from 'Mocks/GetMeQuery';

import { apolloRender, dataHandler, getSnapshot, loadingHandler } from 'testUtils';
import ROUTE from 'Constants/routes';

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const Component = <SettingsMenu />;

let mockClient;
beforeEach(() => {
  mockClient = createMockClient();
});
afterEach(cleanup);

describe('When component is loading', () => {
  it('should render without crashing', () => {
    // Act.
    mockClient.setRequestHandler(GetMeQuery, loadingHandler);
    apolloRender(Component, mockClient);

    // Assert.
    expect(screen.getByTestId('spinner')).toBeInTheDocument();
  });
});

describe('When data is ready', () => {
  beforeEach(() => {
    mockClient.setRequestHandler(GetMeQuery, dataHandler(data));
    apolloRender(Component, mockClient);
  });

  it('should render without crashing', () => {
    expect(screen.getByText(data.me.email, { exact: false })).toBeInTheDocument();
    expect(getSnapshot()).toMatchSnapshot();
  });

  describe('When opening settings', () => {
    // Opens settings
    beforeEach(() => {
      fireEvent.click(screen.getByText(data.me.email, { exact: false }));
    });

    it('should not perform any action when clicking on the separator', () => {
      // Act.
      fireEvent.click(screen.getByText('USER SETTINGS', { exact: false }));

      // Assert.
      expect(screen.getByText('USER SETTINGS', { exact: false })).toBeInTheDocument();
    });

    it('should redirect to ssh key page when clicking on this section', () => {
      // Act.
      fireEvent.click(screen.getByText('SSH KEY', { exact: false }));

      // Assert.
      expect(mockHistoryPush).toHaveBeenCalledTimes(1);
      expect(mockHistoryPush).toHaveBeenCalledWith(ROUTE.USER_SSH_KEY);

      setTimeout(() => {
        expect(screen.queryByText('ssh key', { exact: false })).not.toBeInTheDocument();
      }, 0);
    });
  });
});
