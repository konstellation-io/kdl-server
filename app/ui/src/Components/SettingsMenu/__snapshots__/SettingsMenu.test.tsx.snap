// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When data is ready should render without crashing 1`] = `
<body>
  <div>
    <div
      data-testid="settingsCrumb"
    >
      <div
        class="settings _2rBeV _1z7Bn"
      >
        
        <div
          class="_Z0bIQ"
        >
          <div
            class="_3Loey _BNV4_"
            style="height: 40px; min-height: 40px;"
          >
            admin@konstellation.io
          </div>
          <div
            aria-hidden="true"
            class="rah-static rah-static--height-zero _3bqns"
            style="height: 0px; overflow: hidden;"
          >
            <div
              style="display: none;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
